File: src/io/github/twilight_book/items/ItemUtils.java
package io.github.twilight_book.items;

import io.github.twilight_book.Book;
import io.github.twilight_book.utils.uuid.UUIDTag;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.persistence.PersistentDataContainer;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.*;

public class ItemUtils {
    public enum DamageType {
        PHYSICAL,
        TERRA,
        AER,
        IGNIS,
        AQUA,
        LUMEN,
        UMBRA;
    }

    private static final Map<UUID, ItemInstance> ITEMS = new HashMap<>();
    private static final PersistentDataType<byte[], UUID> UUID_TAG_TYPE = new UUIDTag();
    private static final PersistentDataType<byte[], ItemIdentification> IDENTIFICATION_TAG_TYPE = new IdentificationData();
    private static final List<String> REGISTERED_TAG = Arrays.asList(
            "item",
            "unid"
    );

    public static UUID constructUUID(JavaPlugin plugin, PersistentDataContainer container) {
        NamespacedKey key = new NamespacedKey(plugin, "uuid");

        UUID uuid = container.get(key, UUID_TAG_TYPE);
        if (uuid == null) {
            uuid = UUID.randomUUID();
            container.set(key, UUID_TAG_TYPE, uuid);
        }
        return uuid;
    }

    public static ItemStack createItem(JavaPlugin plugin, ItemInstance inst, String path) {
        YamlConfiguration config = inst.getConfig();

        ItemStack item = new ItemStack(inst.getMaterial(), 1);

        ItemMeta meta = item.getItemMeta();
        if (meta == null) throw new NullPointerException("Somehow, I cannot get the metadata of the item.");

        String ID = config.getString("ID");
        if (ID == null) throw new IllegalArgumentException("Cannot get ID of the item.");

        PersistentDataContainer container = meta.getPersistentDataContainer();
        ITEMS.put(constructUUID(plugin, container), inst);
        container.set(new NamespacedKey(plugin, path), PersistentDataType.STRING, ID);

        if (config.contains("MODEL")) {
            meta.setCustomModelData(config.getInt("MODEL"));
        }

        setIdentification(inst, item);

        item.setItemMeta(meta);
        return item;
    }

    public static ItemInstance getInstByItem(JavaPlugin plugin, ItemStack item, String path) {
        if (item.getType() == Material.AIR) return null;

        ItemMeta meta = item.getItemMeta();
        if (meta == null) return null;

        PersistentDataContainer container = meta.getPersistentDataContainer();

        String ID = container.get(new NamespacedKey(plugin, path), PersistentDataType.STRING);
        if (ID == null) return null;

        UUID uuid = constructUUID(plugin, container);
        if (ITEMS.containsKey(uuid)) {
            return ITEMS.get(uuid);
        } else {
            ItemInstance inst = new ItemInstance(Book.getCfg(), Book.getCfg().getItemByID(ID), getIdentification(item));
            ITEMS.put(uuid, inst);
            return inst;
        }
    }

    public static void setIdentification(ItemInstance inst, ItemStack item){
        ItemMeta meta = item.getItemMeta();
        if (meta == null) throw new NullPointerException("Somehow, I cannot get the metadata of the item.");

        ItemIdentification current = inst.getIdentification();
        if (current == null) return;

        meta.getPersistentDataContainer().set(new NamespacedKey(Book.getInst(), "iden"), IDENTIFICATION_TAG_TYPE, current);
        item.setItemMeta(meta);
    }

    public static ItemIdentification getIdentification(ItemStack item) {
        ItemMeta meta = item.getItemMeta();
        if (meta == null) throw new NullPointerException("Somehow, I cannot get the metadata of the item.");

        return meta.getPersistentDataContainer().get(new NamespacedKey(Book.getInst(), "iden"), IDENTIFICATION_TAG_TYPE);
    }

    public static String getDataTag(JavaPlugin plugin, ItemStack item, String k) {
        ItemMeta meta = item.getItemMeta();
        if (meta == null) throw new NullPointerException("Somehow, I cannot get the metadata of the item.");

        return meta.getPersistentDataContainer().get(new NamespacedKey(plugin, k), PersistentDataType.STRING);
    }

    public static String getItemID(ItemStack item) {
        ItemMeta meta = item.getItemMeta();
        if (meta == null) return null;

        final String[] ID = {null};
        meta.getPersistentDataContainer().getKeys().forEach(
                key -> {
                    if (REGISTERED_TAG.contains(key.getKey())) ID[0] = key.getKey();
                }
        );
        return ID[0];
    }

    public static UUID getUUID(ItemStack item, JavaPlugin plugin) {
        ItemMeta meta = item.getItemMeta();
        if (meta == null) return null;

        return meta.getPersistentDataContainer().get(new NamespacedKey(plugin, "uuid"), UUID_TAG_TYPE);
    }

    public static boolean hasItemID(ItemStack item) {
        ItemMeta meta = item.getItemMeta();
        if (meta == null) return false;

        Set<NamespacedKey> data = meta.getPersistentDataContainer().getKeys();

        for (NamespacedKey key : data) {
            if (key.getNamespace().equals("tsbooklib")) {
                return true;
            }
        }
        return false;
    }

    public static void setDataTag(JavaPlugin plugin, ItemStack item, String k, String v) {
        ItemMeta meta = item.getItemMeta();
        if (meta == null) throw new NullPointerException("Somehow, I cannot get the metadata of the item.");

        meta.getPersistentDataContainer().set(new NamespacedKey(plugin, k), PersistentDataType.STRING, v);
        item.setItemMeta(meta);
    }

    public static void unload() {
        ITEMS.clear();
    }
}

File: src/main/kotlin/io/github/mg138/tsbook/item/ItemUtils.kt
package io.github.mg138.tsbook.item

import io.github.mg138.tsbook.Book
import io.github.mg138.tsbook.attribute.InternalItemType
import io.github.mg138.tsbook.item.data.IdentificationTag
import io.github.mg138.tsbook.item.data.UUIDTag
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.inventory.ItemStack
import org.bukkit.persistence.PersistentDataType
import org.bukkit.plugin.java.JavaPlugin
import java.util.*

object ItemUtils {
    val itemCache: MutableMap<UUID, ItemInstance> = HashMap()

    val itemKey = NamespacedKey(Book.inst, "item")
    val uuidKey = NamespacedKey(Book.inst, "uuid")
    val internalItemTypeKey = NamespacedKey(Book.inst, "internal")
    val itemTypeKey = NamespacedKey(Book.inst, "type")
    val identificationKey = NamespacedKey(Book.inst, "iden")
    val uuidArrayKey = NamespacedKey(Book.inst, "item_uuids")

    inline fun checkItem(item: ItemStack?, actionOnFail: () -> Unit = {}): Boolean {
        if (item == null || item.type == Material.AIR) {
            actionOnFail()
            return false
        }
        return true
    }

    fun createItem(inst: ItemInstance): ItemStack {
        val item = ItemStack(inst.material)
        val meta = item.itemMeta!!
        val uuid = inst.uuid

        val container = meta.persistentDataContainer

        container[uuidKey, UUIDTag] = uuid
        itemCache[uuid] = inst

        val internalItemType = inst.internalItemType
        container[internalItemTypeKey, PersistentDataType.STRING] = internalItemType.string
        container[NamespacedKey(Book.inst, internalItemType.string), PersistentDataType.STRING] = inst.id

        container[itemTypeKey, PersistentDataType.STRING] = inst.itemType.name
        meta.setCustomModelData(inst.model)

        item.itemMeta = meta
        setIdentification(inst, item)
        return item
    }

    fun getInstByItem(plugin: JavaPlugin, item: ItemStack): ItemInstance? {
        if (!checkItem(item)) return null
        val meta = item.itemMeta!!
        val container = meta.persistentDataContainer

        val uuid = container[uuidKey, UUIDTag].let { uuid ->
            itemCache[uuid]?.let { return it }
            UUID.randomUUID()
        }

        val internalItemType = getInternalItemType(item) ?: return null
        val id = container[NamespacedKey(plugin, internalItemType.string), PersistentDataType.STRING] ?: return null

        val inst = ItemInstance(
            id,
            ItemStat.create(
                getIdentification(item),
                id
            ),
            internalItemType,
            uuid
        )
        itemCache[uuid] = inst
        return inst
    }

    private fun setIdentification(inst: ItemInstance, item: ItemStack) {
        val meta = item.itemMeta ?: return
        val identification = inst.itemStat?.identification ?: return
        meta.persistentDataContainer[identificationKey, IdentificationTag] = identification
        item.itemMeta = meta
    }

    private fun getIdentification(item: ItemStack): ItemIdentification? {
        return item.itemMeta!!.persistentDataContainer[identificationKey, IdentificationTag]
    }

    fun getUUID(item: ItemStack): UUID? {
        return item.itemMeta!!.persistentDataContainer[uuidKey, UUIDTag]
    }

    fun getStringTag(item: ItemStack, key: String): String? {
        return item.itemMeta!!.persistentDataContainer[NamespacedKey(Book.inst, key), PersistentDataType.STRING]
    }

    fun getInternalItemType(item: ItemStack): InternalItemType? {
        val container = item.itemMeta!!.persistentDataContainer

        return container[internalItemTypeKey, PersistentDataType.STRING]?.let { InternalItemType.of(it) }
    }

    fun hasItemID(item: ItemStack): Boolean {
        if (!checkItem(item)) return false
        return item.itemMeta!!.persistentDataContainer.has(itemKey, PersistentDataType.STRING)
    }

    fun unload() {
        itemCache.clear()
    }
}